src/theory/partition_generator.cpp

```diff
@@ -133,14 +153,84 @@ TrustNode PartitionGenerator::makeRevisedPartitions()
   else
   {
     vector<Node> nots;
-    for (auto c : d_cubes) nots.push_back(c.notNode());
+    for (const Node& c : d_cubes) nots.push_back(c.notNode());
     Node lemma = NodeManager::currentNM()->mkAnd(nots);
     // Emit not(cube_one) and not(cube_two) and ... and not(cube_n-1)
     emitCube(lemma);
     return stopPartitioning();
   }
 }
 
+
+TrustNode PartitionGenerator::makeFullTrailPartitions()
+{
+  std::vector<TNode> literals = collectDecisionLiterals();
+  uint64_t numVar = static_cast<uint64_t>(log2(d_numPartitions));
+  if (literals.size() >= numVar)
+  {
+    literals.resize(numVar);
+
+    // This complicated thing is basically making a truth table
+    // of with 2^numVar variable so that each row can be emitted as a partition
+    // later. Each entry in resultNodeLists is a row corresponding to a cube:
+    // resultNodeLists = {
+    //   { l1,  l2}
+    //   { l1, !l2}
+    //   {!l1,  l2}
+    //   {!l1, !l2} }
+
+    // total number of cubes/rows
+    size_t total = pow(2, numVar);
+
+    // resultNodeLists is built column by column. 
+    std::vector<std::vector<TNode> > resultNodeLists(total);
+
+    // t is used to determine whether to push the node or its not_node.
+    bool t = false;
+
+    // numConsecutiveTF tracks how many times the node should be consectuively 
+    // true or false in a column.
+    // For example, if numVar=3:
+    // x y z
+    // T T T
+    // T T F
+    // T F T
+    // T F F
+    // F T T
+    // F T F
+    // F F T
+    // F F F
+    // For the first column, numConsecutiveTF = 4, then 2 for the second column, 
+    // and 1 for the third column.
+    size_t numConsecutiveTF = total/2;
+    for (Node n : literals)
+    {
+      Node not_n = n.notNode();
+
+      // loc tracks which row/cube we're on 
+      size_t loc = 0;
+      for (size_t z = 0; z < total / numConsecutiveTF; ++z)
+      {
+        t = !t;
+        for (size_t j = 0; j < numConsecutiveTF; ++j)
+        {
+          resultNodeLists[loc].push_back((t ? n : not_n));
+          ++loc;
+        }
+      }
+
+      numConsecutiveTF = numConsecutiveTF/2;
+    }
+    for (std::vector<TNode> row : resultNodeLists)
```

Comment:
```suggestion
    for (const std::vector<TNode>& row : resultNodeLists)
```
Since you don't want to make a copy of the vector while iterating.